python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_resources.py

```diff
@@ -0,0 +1,171 @@
+# Resource definitions on assets
+#####################################
+
+
+def test_with_resources():
+    @resource
+    def my_resource():
+        ...
+
+    @asset(resource_defs={"foo": my_resource})
+    def asset1():
+        ...
+
+    @asset(required_resource_keys={"foo"})
+    def asset2():
+        ...
+
+    assert asset1 == with_resources([asset2], resource_defs={"foo": my_resource})[0]
+
+
+def test_different_resource_different_asset_keys():
+    @asset(resource_defs={"foo": ResourceDefinition.string_resource()})
+    def asset1(context):
+        assert context.foo == "abc"
+
+    @asset(resource_defs={"bar": ResourceDefinition.string_resource()})
+    def asset2(context):
+        assert context.foo == "xyz"
+
+    assert (
+        AssetGroup([asset1, asset2])
+        .materialize(run_config={"resources": {"foo": {"config": "abc"}, "bar": {"config": "xyz"}}})
+        .success
+    )
+
+
+def test_config_same_resource_multiple_assets():
+    @asset(resource_defs={"foo": ResourceDefinition.string_resource()})
+    def asset1(context):
+        assert context.foo == "abc"
+
+    @asset(resource_defs={"foo": ResourceDefinition.string_resource()})
+    def asset2(context):
+        assert context.foo == "abc"
+
+    assert (
+        AssetGroup([asset1, asset2])
+        .materialize(run_config={"resources": {"foo": {"config": "abc"}}})
+        .success
+    )
+
+
+def test_different_resource_multiple_assets():
+    @resource
+    def resource1():
+        ...
+
+    @resource
+    def resource2():
+        ...
+
+    @asset(resource_defs={"foo": resource1})
+    def asset1(context):
+        ...
+
+    @asset(resource_defs={"foo": resource2})
+    def asset2(context):
+        ...
+
+    group = AssetGroup([asset1, asset2])
+    assert group.materialize(selection=["asset1"]).success
+    assert group.materialize(selection=["asset2"]).success
+    with pytest.raises(
+        match="blah blah different resource definitions for the same resource key. plz use a different resource key."
+    ):
+        group.materialize()
+
+
+def test_io_manager_keys():
+    @io_manager(config_schema=str)
+    def my_configurable_io_manager():
+        ...
+
+    @asset(resource_defs={"foo": my_configurable_io_manager}, io_manager_key="foo")
+    def asset1(context):
+        ...
+
+    @asset(resource_defs={"foo": my_configurable_io_manager}, io_manager_key="foo")
+    def asset2(context):
+        ...
+
+    AssetGroup([asset1, asset2]).materialize(run_config={"resources": {"foo": {"config": "abc"}}})
+
+
+def test_unsatisfied_resource_requirement():
+    @asset(resource_defs={"foo": ResourceDefinition.none_resource()})
+    def asset1(context):
+        ...
+
+    @asset(required_resource_keys={"foo"})
+    def asset2(context):
+        ...
+
+    with pytest.raises(
+        match="blah blah asset2 doesn't have a definition for required resource foo"
+    ):
+        AssetGroup([asset1, asset2]).materialize()
```

Comment:
should you even be able to make an asset group if some assets have required resource keys that aren't satisfied?